### PR TITLE
Verify IPv6 hostnames

### PR DESCRIPTION
**Pull request by yschimke**
Fixes https://github.com/square/okhttp/issues/5885

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/internal/tls/OkHostnameVerifier.kt [hunk 1]
```diff
@@ -51,8 +52,10 @@ object OkHostnameVerifier : HostnameVerifier {
 
   /** Returns true if [certificate] matches [ipAddress]. */
   private fun verifyIpAddress(ipAddress: String, certificate: X509Certificate): Boolean {
+    val canonicalIpAddress = ipAddress.toCanonicalHost()
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/tls/OkHostnameVerifier.kt]: FWIW I didn't consider IP Addresses a host path to optimise too much.  But that's certainly debatable.

File: okhttp/src/test/java/okhttp3/internal/tls/HostnameVerifierTest.java [hunk 2]
```diff
@@ -516,6 +516,42 @@
...
+  @Test public void subjectAltNameWithIPAddresses() throws Exception {
+    // $ cat ./cert.cnf
+    // [req]
+    // distinguished_name=distinguished_name
+    // req_extensions=req_extensions
+    // x509_extensions=x509_extensions
+    // [distinguished_name]
+    // [req_extensions]
+    // [x509_extensions]
+    // subjectAltName=IP:0:0:0:0:0:0:0:1,IP:2a03:2880:f003:c07:face:b00c::2,IP:0::5,IP:192.168.1.1
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/tls/HostnameVerifierTest.java]: Wanna do an IPv4 mapped address also?
https://github.com/square/okhttp/issues/4451

Comment (yschimke) [reply okhttp/src/test/java/okhttp3/internal/tls/HostnameVerifierTest.java]: What do you want the behaviour to be for ::ffff:192.168.1.1?

Comment (yschimke) [reply okhttp/src/test/java/okhttp3/internal/tls/HostnameVerifierTest.java]: Will land that as a follow up.

Comment (swankjesse) [reply okhttp/src/test/java/okhttp3/internal/tls/HostnameVerifierTest.java]: I think it’s the IPv4 address.

Comment (yschimke) [reply okhttp/src/test/java/okhttp3/internal/tls/HostnameVerifierTest.java]: What's the reasoning?  Are you self signing certificates but don't want to specify both forms of IPv4 and IPV6 (mapped IPv4) and you expect it to still work for either?

I'm nervous about Postel's law here.  Maybe keeping it explicit is the better option?

Comment (swankjesse) [reply okhttp/src/test/java/okhttp3/internal/tls/HostnameVerifierTest.java]: The canonical form of certain IPv6 addresses (with colons!) is an IPv4 address (no colons!). It’s weird, but it’s how it’s specified and I’ve come to terms with it.

If we’re canonicalizing IP addresses, we should canonicalize all of ’em and test that it works. I expect the test already passes but I’d like coverage of that case.

Comment (yschimke) [reply okhttp/src/test/java/okhttp3/internal/tls/HostnameVerifierTest.java]: This ok? https://github.com/square/okhttp/pull/5892

**Pull request merged**